### PR TITLE
Parallel Subset Simulation sampler for failure evaluation and optimization

### LINKED ISSUE #19398: Parallel Subset Simulation sampler for failure evaluation and optimization
## Reason
Create an adaptive Monte Carlo sampler class `ParallelSubsetSimulation` to efficiently estimate probabilistic failure of expensive FE models and conduct optimization under uncertainty studies.

## Design
`ParallelSubsetSimulation` sampler class will be created in stochastic tools. In addition, `AdaptiveMonteCarloDecision` and `AdaptiveMonteCarloUtils` classes will be modified to accommodate `ParallelSubsetSimulation`. Corresponding docs and tests will also be created.

## Impact
This development of the above class should not impact the behavior of existing objects.

### PR DESCRIPTION
**Pull request by somu15**
## Reason
Create an adaptive Monte Carlo sampler class `ParallelSubsetSimulation` to efficiently estimate probabilistic failure of expensive FE models and conduct optimization under uncertainty studies.

## Design
`ParallelSubsetSimulation` sampler class will be created in stochastic tools. In addition, `AdaptiveMonteCarloDecision` and `AdaptiveMonteCarloUtils` classes will be modified to accommodate `ParallelSubsetSimulation`. Corresponding docs and tests will also be created.

## Impact
This development of the above class should not impact the behavior of existing objects.

Closes #19398

### CODE REVIEW

**Comment by somu15** [comment]
Not sure why `test:finite_volume/wcns/materials.transient` is exodiffing.

**Comment by grmnptr** [comment]
Do you mind if I also review this one? I am interested in the technique and would love to learn about the implementation.

**Comment by zachmprince** [comment]
@grmnptr I wouldn't review it right now. I gave some changes to Som and he is still trying to make sure it didn't break anything. It would be great to get your input though!

**Comment by somu15** [comment]
@zachmprince sorry I couldn't test it on Triso this week. I will work on this next week.

**Review by grmnptr — CHANGES_REQUESTED** [review]
Very nice piece of work! I added a few minor suggestions/comments, just in case.

File: modules/stochastic_tools/src/samplers/ParallelSubsetSimulation.C [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  InputParameters params = Sampler::validParams();
+  params.addClassDescription("Parallel Subset Simulation sampler.");
+  params.addRequiredParam<std::vector<DistributionName>>(
+      "distributions",
+      "The distribution names to be sampled, the number of distributions provided defines the "
+      "number of columns per matrix.");
+  params.addRequiredParam<ReporterName>("output_reporter",
+                                        "Reporter with results of samples created by trainer.");
+  params.addRequiredParam<ReporterName>("inputs_reporter", "Reporter with input parameters.");
+  params.addRangeCheckedParam<Real>("subset_probability",
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/src/samplers/ParallelSubsetSimulation.C:29]: This param is range checked. So, it's automatically tested right?

File: modules/stochastic_tools/doc/content/source/samplers/ParallelSubsetSimulation.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
+# Parallel Subset Simulation (PSS)
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/doc/content/source/samplers/ParallelSubsetSimulation.md:1]: Yes, this is a great idea! Right now, this class is described as being used for failure analysis and stochastic optimization. We are using it for stochastic optimization for a different numerical model. But yes, this sampler is a variance reduction method.

File: modules/stochastic_tools/src/reporters/AdaptiveMonteCarloDecision.C [hunk 3]
```diff
@@ -34,26 +36,43 @@ AdaptiveMonteCarloDecision::AdaptiveMonteCarloDecision(const InputParameters & p
...
     _step(getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")->timeStep()),
     _sampler(getSampler("sampler")),
     _ais(dynamic_cast<const AdaptiveImportanceSampler *>(&_sampler)),
+    _pss(dynamic_cast<const ParallelSubsetSimulation *>(&_sampler)),
     _check_step(std::numeric_limits<int>::max())
 {
 
   // Check whether the selected sampler is an adaptive sampler or not
-  if (!_ais)
+  if (!_ais && !_pss)
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/src/reporters/AdaptiveMonteCarloDecision.C:48]: Done!

File: modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C [hunk 4]
```diff
@@ -32,7 +34,6 @@ computeSTD(const std::vector<Real> & data, const unsigned int & start_index)
   }
```
[the remark this answers is not in the document]
Comment (somu15) [reply modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C:34]: Yes, as you mentioned, practically, there wouldn't be much difference between N and N-1. Unless, the user request very few samples per subset, like 10.

**Review by zachmprince — APPROVED** [review]

**Pull request merged**